frontend/gateway/container.go

```diff
@@ -0,0 +1,242 @@
+package gateway
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	"github.com/moby/buildkit/cache"
+	"github.com/moby/buildkit/executor"
+	"github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/solver"
+	opspb "github.com/moby/buildkit/solver/pb"
+	"github.com/moby/buildkit/util/stack"
+	utilsystem "github.com/moby/buildkit/util/system"
+	"github.com/moby/buildkit/worker"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+type Container interface {
+	client.Container
+	// OnRelease allows callbacks to free up resources when the container exits.
+	// The functions are called in LIFO order.
+	OnRelease(func() error)
```

Comment:
Not sure I understand the need for this. Why doesn't `Wait()` just return if container was released?

Reply:
We only have a `Wait` on the processes, not the container as a whole:
```
type Container interface {
	Start(context.Context, StartRequest) (ContainerProcess, error)
	Release(context.Context) error
}
```

I can add a `Wait() error` to the container interface which I think which could serve the same purpose as my `OnRelease`:
```
go func() {
  defer releaseStuff()
  ctr.Wait()
}()
```

Reply:
I have updated to remove the unnecessary interface.